patroni/postgresql.py

```diff
@@ -655,12 +750,84 @@ def reload(self):
             self.call_nowait(ACTION_ON_RELOAD)
         return ret
 
-    def restart(self):
-        self.set_state('restarting')
-        ret = self.stop(block_callbacks=True) and self.start(block_callbacks=True)
-        if ret:
-            self.call_nowait(ACTION_ON_RESTART)
+    def check_for_startup(self):
+        """Checks PostgreSQL status and returns if PostgreSQL is in the middle of startup."""
+        if self.is_starting():
+            return not self.check_startup_state_changed()
         else:
+            ready = self.pg_isready()
```

Comment:
> if the system is so slow that pg_ctl status makes ttl run out, then losing the master key is the desired outcome as the system is not really fit to be running a database

You are right, but from another side there is not guaranty that the new master will do better... One missing index can ruin everything :)

> we can't work like this and shutting everything down in the hope that someone else is in a better position.

Usually all database instances have more or less the same capacity and chances that situation will improve are quite small. Moreover, probably it will case not a single failover but bunch of them in a row...

> Do you want to take a shot at moving pg_isready -> state=starting after a failed is_leader check or should I take it on?

I will try to work on it.